crates/lace-bin/src/typechecking/error.rs

```diff
@@ -1,29 +1,346 @@
 use crate::parser::ast::{BinaryOp, UnaryOp};
 use crate::parser::span::Span;
-use miette::Diagnostic;
+use miette::{Diagnostic, LabeledSpan};
 use thiserror::Error;
 
-#[derive(Debug, Error, PartialEq, Clone, Diagnostic)]
-#[diagnostic()]
+#[derive(Debug, PartialEq, Clone)]
+pub struct FailedUnification {
+    pub expected: String,
+    pub was: String,
+    pub was_span: Span,
+}
+
+pub trait ResultExt {
+    fn on_failed_unification(self, f: impl FnOnce(FailedUnification) -> TypeError) -> Self;
```

Comment:
no, essentially failed unifications invert the error structure. This happens when a function doesn't quite know why an error occurred, but it did because of whichever type was passed in. Where we pass in a concrete type, we then catch it with on_failed_unification because there we know the context why an error did occur.

Reply:
So it's the same as adding error context but then only keeping the inner context?